kobo/apps/mass_emails/tasks.py

```diff
@@ -0,0 +1,122 @@
+from datetime import date
+from math import ceil
+
+from django.conf import settings
+from django.core.cache import cache
+from django.db.models import Count, Q
+from django.utils.translation import gettext
+
+from kobo.celery import celery_app
+from kpi.exceptions import ExecutionBlockedException
+from kpi.utils.log import logging
+from kpi.utils.mailer import EmailMessage, Mailer
+from .models import (
+    USER_QUERIES,
+    EmailStatus,
+    MassEmailConfig,
+    MassEmailJob,
+    MassEmailRecord,
+)
+
+templates_placeholders = {
+    '##username##': 'username',
+    '##full_name##': 'full_name',
+    '##plan_name##': 'plan_name',
+}
+
+
+def render_template(template, data):
+    rendered = template
+    for placeholder, value in templates_placeholders.items():
+        rendered = rendered.replace(placeholder, data[value])
+    return rendered
+
+
+def create_job(email_config):
+    logging.info(f'Creating job for {email_config=}')
+    job = MassEmailJob.objects.create(
+        email_config=email_config,
+    )
+    users = USER_QUERIES[email_config.query]()
+    records = [
+        MassEmailRecord(
+            user=user,
+            email_job=job,
+            status=EmailStatus.ENQUEUED,
+        )
+        for user in users
+    ]
+    MassEmailRecord.objects.bulk_create(records)
+
+
+@celery_app.task
+def send_emails(email_config_uid: str, should_create_job: bool = False):
+    today = date.today().isoformat()
+    cache_key = f'MassEmailConfig_Lock_{today}_{email_config_uid}'
+    if cache.get(cache_key) == 'Locked':
+        raise ExecutionBlockedException
+    cache.set(cache_key, 'Locked', 60*60*24) # 1 day lock
```

Comment:
I think we can simplify this a bit without having to use the cache. We can have something like
```
current_running_jobs = my_config.jobs.filter(Q(jobs__records__status=EmailStatus.ENQUEUED))
if current_running_jobs > 0:
    raise NotGonnaHappen('Send is already in progress')
...otherwise, create a new job and go on our way
```
we have a task to cleanup records that have been enqueued for too long, so even if whatever called send_email before is no longer running, we should still get unblocked eventually